core/src/main/java/org/apache/iceberg/TableProperties.java

```diff
@@ -134,4 +134,10 @@ private TableProperties() {
 
   public static final String ENGINE_HIVE_ENABLED = "engine.hive.enabled";
   public static final boolean ENGINE_HIVE_ENABLED_DEFAULT = false;
+
+  public static final String WRITE_ISOLATION_LEVEL = "write.isolation-level";
```

Comment:
@rdblue, naming is still TBD.

Reply:
My current thinking on naming here is that all query engines use the same API to commit. Right now, we only have appends that guarantee serializable isolation. I think query engines supporting update/delete/merge should respect this property or throw an exception if they cannot satisfy the required isolation level.

We may add a property specific to query engines if we think we will need different defaults. I'd say those will be optional. If a query engine specific value is not set, use this for all engines.

Reply:
My only reservation about this is that the isolation level varies by operation. Dynamic partition replacement doesn't check whether the partitions that will be replaced have been modified since the query started. Should we go back and update those operations?

I think people also expect those operations to have the semantics that they currently do because that matches Hive. It would be strange for them to either implement a stricter isolation level, or to prevent using a stricter isolation level in merge operations.

Reply:
Dynamic partition overwrite does mess up the situation. I like the `row-level` namespace we introduced for properties. We could make it `write.row-level.isolation` to match `write.row-level.mode`. That way, it is clear it applies only to update/delete/merge.

If we decide to have engine specific defaults later, we can introduce `write.row-level.spark.isolation` or `engine.spark.write.row-level.isolation`. The latter can serve as a basis for providing any engine specific properties..

Reply:
Since we are using "merge" in the code, what do you think about `write.merge.isolation` instead of `row-level`?

Reply:
I think having only `write.merge.isolation` may not be descriptive enough since users won't know that we treat delete as a simplified version of merge.

I would keep `write.row-level.isolation` with `write.row-level.mode` or introduce a pair of configs for each operation type like `write.delete.isolation` and `write.delete.mode`. I am ok with both options.